### ISSUE TITLE
Wrong code dimming when asserting comparisons of sys.version_info and tuples

### ISSUE THREAD

**Issue by detuur** [microsoft/pyright#7132]
## Environment

Pylance version: v2023.12.1
Python version: 3.10.12
OS Version: Ubuntu 22.04 LTS

## Issue

When using code patterns like `assert sys.version_info > (3,10)` to gate functionality in code, Pylance will perform the tuple comparison wrongly, and will dim code that is actually reachable. See a visual example in my screenshot below.

![image](https://github.com/microsoft/pylance-release/assets/2746724/e48a2a54-f0de-47d3-a12e-6ed29264bedc)

<details>
<summary>Code in text form</summary>

    ## Testing dimming of unreachable code (Python 3.10.12)
    
    # In this function, unreadable code is clearly dimmed
    def test_sane():
        pass
        return
        pass
    
    # Tuple comparisons do not elicit dimming
    def test_dimming_with_tuple_assertions():
        assert (1,0)   >  (0,1)
        assert (1,0)   >= (1,0)
        assert (1,0,1) >= (1,0)
        assert (1,0,1) >  (1,0)
        print("reachable, because all these tuple comparisons are True")
    
        assert (2) > (3)
        print("unreachable, but not dimmed")
    
        assert False
        print("other types of assertions are dimmed")
    
    def test_dimming_with_version_assertions():
        import sys
        print("reachable for any python version")
    
        assert sys.version_info > (3,9)
        print("reachable for python > 3.9")
    
        assert sys.version_info >= (3,10)
        print("reachable for python >= 3.10")
    
        assert sys.version_info > (3,10)
        # Code starting here is dimmed, despite the fact that it is reachable
        print("reachable for python > 3.10")
    
        assert sys.version_info >= (3,11)
        print("reachable for python >= 3.11")
    
    def do_runs():
        try:
            test_dimming_with_tuple_assertions()
        except:
            pass
        try:
            test_dimming_with_version_assertions()
        except:
            pass
    
    do_runs()
    print("..done")
</details>

## Extra info
No logs or traces were captured, doesn't seem applicable to this issue. Ping me if you'd like extra information.

**Comment by detuur** [comment]
Yes I see how that can help. It's in there now, under the screenshot.

**Comment by erictraut** [comment]
I'm not able to repro a problem. Pyright appears to be working as intended.

The `sys.version_info` comparisons are evaluated specially by a type checker. This behavior is dictated by the Python typing spec [here](https://typing.readthedocs.io/en/latest/spec/directives.html#version-and-platform-checking).

Pyright statically evaluates these statements based on the current `pythonVersion` you've specified in your [pyright configuration](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options). If you don't specify a `pythonVersion`, pyright (and pylance) default to the version of the currently-selected Python interpreter. For example, if you're running a Python 3.10 interpreter, `pythonVersion` defaults to 3.10, and the `sys.version_info` statements will be evaluated accordingly. If you would like pyright to assume some other version, you can specify this in your pyrightConfig.json or pyproject.toml file.

I'm going to close this issue because pyright is working as designed.

**Comment by detuur** [comment]
The discrepancy is not about the version mismatch. The discrepancy is in how Pyright compares the `version_info` tuple, which is a 5-tuple, against (1..4)-tuples.

On my system, `sys.version_info` is `sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)`.

This is how the comparisons go on my system:
```python3
sys.version_info > (3) # True
sys.version_info > (3,10) # True
sys.version_info > (3,10,12) # True
sys.version_info > (3,10,12,'final') # True
sys.version_info > (3,10,12,'final',0) # False
```
Only the last comparison (5-tuple vs 5-tuple) is false. In every other comparison, Python treats the `>` operator as a `>=` operator.

**Comment by erictraut** [comment]
Ah, I see what you mean. Thanks for the added details.

Pyright is conforming to the Python typing spec here. The spec requires type checkers to honor comparison expressions that involve one or two tuple elements in the tuple. When you configure your type checker (whether it's pyright, mypy, pyre, etc.), you specify the python version you would like it to assume when performing static analysis. That version is always specified as a two-element form like `3.10` or `3.11`. If your code include additional elements in a `sys.version_info` comparison, it's not clear how a type checker should interpret the additional elements beyond the second. It's indeterminate and ambiguous since the configuration version supports only two elements.

Pyright's current implementation simply ignores any elements after the second in the code, which I think is defensible, although it can admittedly produce the wrong answer some of the time. The other approach is to assume that if there are more than two elements, the statement will always evaluate to False. That's also defensible, but it will likewise produce the wrong answer some of the time.

For comparison, here's how pyright statically evaluates these statements versus mypy. As you can see, they disagree with each other and with the actual runtime behavior.

```python
# Pyright static evaluations
sys.version_info > (3,) # True
sys.version_info > (3,10) # True
sys.version_info > (3,10,12) # True
sys.version_info > (3,10,12,'final') # True
sys.version_info > (3,10,12,'final',0) # True

# Mypy static evaluations
sys.version_info > (3,) # True
sys.version_info > (3,10) # False
sys.version_info > (3,10,12) # True
sys.version_info > (3,10,12,'final') # True
sys.version_info > (3,10,12,'final',0) # True
```

I think pyright's current behavior is correct in terms of the typing spec. It's also working as designed, so I don't consider this a "bug", but I can understand how you would see the current behavior as suboptimal.

Maybe the best answer is to change pyright's configuration mechanism to allow `pythonVersion` to be specified using the full five elements. Then it could evaluate all of these forms statically without any ambiguity.

I'm going to re-open the ticket and change it from a bug report to a feature request. I'll give it more thought about whether it's worth implementing this feature.

**Comment by detuur** [comment]
Hi Eric,

Thank you for the elaborate reply. I genuinely appreciate the time you've set aside for this. I fear there might be a repro issue though, because this is not the way the evaluations are happening in VSCode for me. On my side, only the 1-tuple and 5-tuple cases are correctly evaluated. I've created a different test case, which gives the following: 

(Python, Pylance, and Black are the only extensions enabled.)
![image](https://github.com/microsoft/pyright/assets/2746724/06a54958-1f3c-4290-9d38-110bf73e818b)

<details>
<summary>Code used</summary>

    import sys
    
    # To illustrate that pyright sees version_info as (3,10)
    def matching_tuple():
        assert sys.version_info == (3,10)
        pass
        # True for only the (3,10) 2-tuple. For (3..5)-tuples,
        # only true when they start with (3,10, ...)
    
    # Correct behaviour, reachable code is undimmed.
    def one_tuple():
        assert sys.version_info > (3,)
        pass # True
    
    # Misleading behaviour. Even by its own understanding of
    # the current version_info, it is dimming code that is
    # reachable.
    def two_tuple():
        assert sys.version_info > (3,10)
        pass # False
    
    # Also misleading, but not as impactful as the 2-tuple case.
    def three_tuple():
        assert sys.version_info > (3,10,12)
        pass # False
    
    # Same as 3-tuple case.
    def four_tuple():
        assert sys.version_info > (3,10,12,'final')
        pass # False
    
    # Correct behaviour again.
    def five_tuple():
        assert sys.version_info > (3,10,12,'final', 0)
        pass # False
</details>

I'm mostly bothered about the 2-tuple case, as that's probably the most common case for static analysis. It's also occurred to me that this might not exactly be an issue in Pyright, but in the way it's implemented by Pylance downstream. Either way, LMK if you need extra info.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.355, which I just published.
